src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c

```diff
@@ -2333,8 +2334,27 @@ static int __icap_download_bitstream_axlf(struct platform_device *pdev,
 		num_dev = xocl_fdt_parse_blob(xdev, icap->partition_metadata,
 				icap_get_section_size(icap, PARTITION_METADATA),
 				&subdevs);
-		for (num_dev--; num_dev >= 0; num_dev--)
-			xocl_subdev_create(xdev, &subdevs[num_dev].info);
+		ICAP_INFO(icap, "found %d sub devices", num_dev);
+		for (i = 0; i < num_dev; i++)
+			xocl_subdev_create(xdev, &subdevs[i].info);
+	}
+
+	if (num_dev > 0) {
+		for (i = 0; i < ICAP_MAX_NUM_CLOCKS; i++) {
+			if (icap->icap_clock_bases[i])
+				break;
+		}
+		xocl_subdev_create_by_level(xdev, XOCL_SUBDEV_LEVEL_URP);
```

Comment:
should 2347,2348 be within 2334-2339? So that xocl_cmc_free can be added outside (after) this if(num_dev > 0) {}.